File: js/Assignments/utils/plagiarismChecker.js
const fs = require('fs');
const path = require('path');

function calculateSimilarity(str1, str2) {
    // Remove whitespace and convert to lowercase for better comparison
    str1 = str1.replace(/\s+/g, '').toLowerCase();
    str2 = str2.replace(/\s+/g, '').toLowerCase();
    
    if (str1 === str2) return 100;
    if (str1.length === 0 || str2.length === 0) return 0;

    // Using Levenshtein distance for similarity
    const matrix = Array(str2.length + 1).fill().map(() => Array(str1.length + 1).fill(0));

    for (let i = 0; i <= str1.length; i++) matrix[0][i] = i;
    for (let j = 0; j <= str2.length; j++) matrix[j][0] = j;

    for (let j = 1; j <= str2.length; j++) {
        for (let i = 1; i <= str1.length; i++) {
            if (str1[i - 1] === str2[j - 1]) {
                matrix[j][i] = matrix[j - 1][i - 1];
            } else {
                matrix[j][i] = Math.min(
                    matrix[j - 1][i - 1] + 1,
                    matrix[j][i - 1] + 1,
                    matrix[j - 1][i] + 1
                );
            }
        }
    }

    const distance = matrix[str2.length][str1.length];
    const maxLength = Math.max(str1.length, str2.length);
    return Math.round((1 - distance / maxLength) * 100);
}

function checkPlagiarism(submissionPath, submissionsDir) {
    const results = [];
    const submissionContent = fs.readFileSync(submissionPath, 'utf8');
    const submissions = fs.readdirSync(submissionsDir);

    for (const otherSubmission of submissions) {
        const otherPath = path.join(submissionsDir, otherSubmission);
        
        // Skip if it's a directory or the same file
        if (!fs.statSync(otherPath).isFile() || otherPath === submissionPath) continue;
        
        const otherContent = fs.readFileSync(otherPath, 'utf8');
        const similarity = calculateSimilarity(submissionContent, otherContent);

        if (similarity > 80) { // Threshold for similarity
            results.push({
                file: otherSubmission,
                similarity
            });
        }
    }

    return results;
}

module.exports = { checkPlagiarism }; 

File: js/Assignments/utils/plagiarism_checker.py
import os
from difflib import SequenceMatcher
import fnmatch

def calculate_similarity(text1, text2):
    """Calculate similarity ratio between two texts using SequenceMatcher."""
    # Remove whitespace and convert to lowercase for better comparison
    text1 = ''.join(text1.split()).lower()
    text2 = ''.join(text2.split()).lower()
    
    return SequenceMatcher(None, text1, text2).ratio() * 100

def check_plagiarism(submission_path, submissions_dir, file_patterns=['*.py', '*.js', '*.html', '*.css']):
    """
    Check submission against other submissions for plagiarism.
    Returns a list of files with similarity above threshold.
    """
    results = []
    
    # Read the submission content
    with open(submission_path, 'r', encoding='utf-8') as f:
        submission_content = f.read()
    
    # Get all files in submissions directory
    for root, _, files in os.walk(submissions_dir):
        for file in files:
            # Check if file matches any of the patterns we're looking for
            if any(fnmatch.fnmatch(file, pattern) for pattern in file_patterns):
                file_path = os.path.join(root, file)
                
                # Skip if it's the same file
                if os.path.samefile(file_path, submission_path):
                    continue
                
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        other_content = f.read()
                        
                    similarity = calculate_similarity(submission_content, other_content)
                    
                    # If similarity is above 90%, add to results
                    if similarity > 90:
                        results.append({
                            'file': os.path.relpath(file_path, submissions_dir),
                            'similarity': round(similarity, 2)
                        })
                except Exception as e:
                    print(f"Error reading file {file_path}: {str(e)}")
    
    return results

if __name__ == "__main__":
    # Example usage
    submission = "path/to/submission/file.py"
    submissions_dir = "Assignments/Submissions"
    results = check_plagiarism(submission, submissions_dir)
    print(results) 